### PR TITLE
fixes #217 locking issues

### LINKED ISSUE #217: File lock issues with version 0.5.2
I've encountered an issue with the project locking process as per the below.

```
[2023-03-30, 09:11:08 BST] {taskinstance.py:1768} ERROR - Task failed with exception
Traceback (most recent call last):
  File "/home/airflow/.local/lib/python3.10/site-packages/cosmos/providers/dbt/core/operators.py", line 387, in execute
    result = self.build_and_run_cmd(context=context)
  File "/home/airflow/.local/lib/python3.10/site-packages/cosmos/providers/dbt/core/operators.py", line 312, in build_and_run_cmd
    return self.run_command(cmd=dbt_cmd, env=env)
  File "/home/airflow/.local/lib/python3.10/site-packages/cosmos/providers/dbt/core/operators.py", line 276, in run_command
    shutil.copytree(self.project_dir, target_dir, ignore=exclude)
  File "/usr/local/lib/python3.10/shutil.py", line 559, in copytree
    return _copytree(entries=entries, src=src, dst=dst, symlinks=symlinks,
  File "/usr/local/lib/python3.10/shutil.py", line 513, in _copytree
    raise Error(errors)
shutil.Error: [('/opt/***/dbt/community_pharmacy/snapshots', '/tmp/dbt/community_pharmacy/snapshots', "[Errno 17] File exists: '/tmp/dbt/community_pharmacy/snapshots'"), ('/opt/***/dbt/community_pharmacy/tests', '/tmp/dbt/community_pharmacy/tests', "[Errno 17] File exists: '/tmp/dbt/community_pharmacy/tests'"), ('/opt/***/dbt/community_pharmacy/models', '/tmp/dbt/community_pharmacy/models', "[Errno 17] File exists: '/tmp/dbt/community_pharmacy/models'"), ('/opt/***/dbt/community_pharmacy/macros', '/tmp/dbt/community_pharmacy/macros', "[Errno 17] File exists: '/tmp/dbt/community_pharmacy/macros'"), ('/opt/***/dbt/community_pharmacy/seeds', '/tmp/dbt/community_pharmacy/seeds', "[Errno 17] File exists: '/tmp/dbt/community_pharmacy/seeds'"), ('/opt/***/dbt/community_pharmacy/analyses', '/tmp/dbt/community_pharmacy/analyses', "[Errno 17] File exists: '/tmp/dbt/community_pharmacy/analyses'")]
```

My suspicion is that it's due to this [line](https://github.com/astronomer/astronomer-cosmos/blob/astronomer-cosmos-v0.5.2/cosmos/providers/dbt/core/operators.py#L275)
If we remove the whole directory then surely we're also removing the lock file that's there to stop the other tasks from performing the same activity?

### PR DESCRIPTION
**Pull request by MikeWallis42**
fixes #217

I wasn't entirely sure on the need for 2 different locking mechanisms so I've attempted to simplify the process.
I've also put this in to a seperate method so that we can test it independently of executing the dbt logic.
Unit tests provided although I'm skeptical about wether the threaded test really works due to the size of the test project added.

Also added a hatch test env for local testing purposes, similar to the docs one.

### CODE REVIEW

**Review by pohek321 — CHANGES_REQUESTED** [review]
@MikeWallis42 thank you for submitting this! I did run into one error while testing. LMK if the suggestion I provided was contextually accurate!

File: cosmos/providers/dbt/core/operators.py [hunk 1]
```diff
@@ -308,6 +304,7 @@ def build_and_run_cmd(
         dbt_cmd.extend(["--profile", profile])
         # set env vars
         env = self.get_env(context, profile_vars)
+        self.sync_temp_project()
 
         return self.run_command(cmd=dbt_cmd, env=env)
```
[the remark this answers is not in the document]
Comment (MikeWallis42) [reply cosmos/providers/dbt/core/operators.py]: @pohek321 sorry I completely missed that! all updated now, thanks for spotting it.

File: cosmos/providers/dbt/core/operators.py [hunk 2]
```diff
@@ -215,75 +211,75 @@ def add_global_flags(self) -> list[str]:
...
+                for dir_object in os.listdir(target_path):
+                    dir_object: str
+                    child_path = target_path.joinpath(dir_object)
+                    if child_path.is_dir():
+                        if dir_object not in ["logs", "target"]:
+                            shutil.rmtree(child_path)
+                    elif dir_object not in [".lock"]:
+                        os.remove(child_path)
+                shutil.copytree(
+                    source_path, target_path, ignore=exclude, dirs_exist_ok=True
```
Comment (jbandoro) [inline cosmos/providers/dbt/core/operators.py:260]: hey just wanted to comment here that the `dirs_exist_ok` arg is not available in python 3.7, so if it's required the minimum python version for cosmos in the pyproject [here](https://github.com/astronomer/astronomer-cosmos/blob/b8fbe4c463725e7bbc0ef29b4bfecfb6c0295b2f/pyproject.toml#L11) should be 3.8

Comment (MikeWallis42) [reply cosmos/providers/dbt/core/operators.py:260]: Thanks @jbandoro, I've adjusted the test environment to be python 3.7 so it supports the minimum version so hopefully I don't make that same mistake again!
Just about to push up a commit that makes it 3.7 compliant.

**Comment by jlaneve** [comment]
@MikeWallis42 I think we're going to remove the file syncing in favor of just a new tmp dir each time - check out [this PR](https://github.com/astronomer/astronomer-cosmos/pull/233). What do you think?

**Comment by MikeWallis42** [comment]
@jlaneve, sorry I'm a little late to the party, I'll go take a look now, I know it's already merged but still. I guess this is the problem with me being in Europe 😄

**Pull request closed without merging**